ccip.go

```diff
@@ -0,0 +1,163 @@
+package ens
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/wealdtech/go-ens/v3/contracts/offchainresolver"
+	"github.com/wealdtech/go-ens/v3/contracts/universalresolver"
+)
+
+func getCcipReadError(err error) (bool, string) {
+	var jsonErr, ok = err.(rpc.DataError)
+	if !ok {
+		return false, ""
+	}
+	errData, ok := jsonErr.ErrorData().(string)
+	if !ok {
+		return false, ""
+	}
+	return (len(errData) >= 10 && errData[:10] == offchainLookupSignature), errData
+}
+
+func CCIPRead(backend bind.ContractBackend, rAddr common.Address, revertData string) ([]byte, error) {
+	hexBytes, err := hex.DecodeString(revertData[2:])
+	if err != nil {
+		return nil, err
+	}
+
+	uAbi, err := universalresolver.ContractMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
+	// Extracting error details from the revert data
+	var sig [4]byte
+	copy(sig[:], hexBytes[:4])
+	abiErr, err := uAbi.ErrorByID(sig)
+	if err != nil {
+		return nil, err
+	}
+	errArgs, err := abiErr.Inputs.Unpack(hexBytes[4:])
```

Comment:
Need a check here that the revert data is more than 4 bytes in length to avoid a potential panic.